File: UWP/ViewModels/User/UserViewModel.cs
using System;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Toolkit.Uwp.Extensions;
using PropertyChanged;
using WeiPo.Common;
using WeiPo.Services;
using WeiPo.Services.Models;

namespace WeiPo.ViewModels.User
{
    public class UserViewModel : ViewModelBase
    {
        public UserViewModel(UserModel user, Action updatePivot)
        {
            UpdatePivot = updatePivot;
            Init(user.Id);
        }

        public UserViewModel(string name, Action updatePivot)
        {
            UpdatePivot = updatePivot;
            Init(name);
        }

        public UserViewModel(long id, Action updatePivot)
        {
            UpdatePivot = updatePivot;
            Init(id);
        }

        public bool IsLoading { get; private set; }

        public ProfileData Profile { get; private set; }

        public ProfileData Me { get; private set; }

        [DependsOn(nameof(Profile))]
        public Services.Models.Tab[] Tabs
        {
            get
            {
                return Profile?.TabsInfo?.Tabs?.Concat(new[]
                {
                    new Services.Models.Tab
                    {
                        Title = "Follow".GetLocalized(),
                        TabType = "follow",
                        Containerid = Profile.UserInfo.Id.ToString()
                    },
                    new Services.Models.Tab
                    {
                        Title = "Fans".GetLocalized(),
                        TabType = "fans",
                        Containerid = Profile.UserInfo.Id.ToString()
                    }
                }).ToArray();
            }
        }

        [DependsOn(nameof(Profile))]
        public bool IsNotMe => Profile?.UserInfo?.Id != Me?.UserInfo?.Id;

        public Action UpdatePivot { get; }

        public async Task UpdateFollowState()
        {
            if (IsLoading)
            {
                return;
            }
            IsLoading = true;
            UserModel model;
            if (Profile.UserInfo.Following)
            {
                model = await Singleton<Api>.Instance.Unfollow(Profile.UserInfo.Id);
            }
            else
            {
                model = await Singleton<Api>.Instance.Follow(Profile.UserInfo.Id);
            }
            Profile.UserInfo.Following = !Profile.UserInfo.Following;
            OnPropertyChanged(nameof(Profile));
            IsLoading = false;
        }

        private async void Init(long id)
        {
            if (IsLoading)
            {
                return;
            }

            IsLoading = true;
            await InitProfile(id);
            IsLoading = false;
        }

        private async void Init(string name)
        {
            if (IsLoading)
            {
                return;
            }

            IsLoading = true;
            var id = await Singleton<Api>.Instance.UserId(name);
            await InitProfile(id);
            IsLoading = false;
        }

        private async Task InitProfile(long id)
        {
            Me = await Singleton<Api>.Instance.Me();
            Profile = await Singleton<Api>.Instance.Profile(id);
            UpdatePivot.Invoke();
        }

    }
}

File: Android/app/src/main/java/moe/tlaster/weipo/viewmodel/UserViewModel.kt
package moe.tlaster.weipo.viewmodel

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import moe.tlaster.weipo.common.extensions.async
import moe.tlaster.weipo.common.extensions.runOnMainThread
import moe.tlaster.weipo.services.Api
import moe.tlaster.weipo.services.models.Config
import moe.tlaster.weipo.services.models.ProfileData

open class UserViewModel : ViewModel() {
    lateinit var config: Config
    val profile = MutableLiveData<ProfileData>()

    fun updateFollow() {
        profile.value?.userInfo?.following?.let { state ->
            profile.value?.userInfo?.id?.let { id ->
                async {
                    val result = if (state) {
                        Api.unfollow(id)
                    } else {
                        Api.follow(id)
                    }
                    val value = Api.profile(id)
                    runOnMainThread {
                        profile.value = value
                    }
                }
            }
        }
    }

    fun initProfile(name: String) {
        async {
            val id = Api.userId(name)
            initProfile(id)
        }
    }

    fun initProfile(id: Long) {
        async {
            config = Api.config()
            val value = Api.profile(id)
            runOnMainThread {
                profile.value = value
            }
        }
    }

    fun initMe() {
        async {
            config = Api.config()
            config.uid?.toLongOrNull().takeIf {
                it != null && it != 0L
            }?.let {
                Api.profile(it)
            }?.let {
                runOnMainThread {
                    profile.value = it
                }
            }
        }
    }
}